clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -244,6 +267,50 @@ public void forceClose() {
         initiateClose();
     }
 
+    private void getPID(long remainingTime) {
+        // Send an InitPIDRequest, wait for the response, retrieve the PID from the response if success. Return it.
+        // Otherwise throw exception.
+        // Pick a random broker to send the initPIDRequest to.
+        metadata.requestUpdate();
+        long start = time.milliseconds();
+        Node node = client.leastLoadedNode(time.milliseconds());
+        while (!client.ready(node, time.milliseconds()) && 0 < remainingTime) {
```

Comment:
It might make sense to check for disconnects. If the node is disconnected, we can shortcut return and pick another node.